client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/coverage/methodreplacement/thirdpartyclasses/GsonClassReplacement.java

```diff
@@ -7,44 +7,144 @@
         return "com.google.gson.Gson";
     }
 
-    // TODO all versions of fromJson
-
     @Replacement(replacingStatic = false,
             type = ReplacementType.TRACKER,
             id = "fromJson_string_class",
             usageFilter = UsageFilter.ANY,
-            category = ReplacementCategory.BASE)
-    public static Object fromJson(Object caller, String json, Class<?> classOfT){
+            category = ReplacementCategory.EXT_0)
+    public static Object fromJson(Object caller, String json, Class<?> classOfT) {
 
-        if(caller == null){
+        if (caller == null) {
             throw new NullPointerException();
         }
 
         ClassToSchema.registerSchemaIfNeeded(classOfT);
-        JsonTaint.handlePossibleJsonTaint(json,classOfT);
+        JsonTaint.handlePossibleJsonTaint(json, classOfT);
 
         Method original = getOriginal(singleton, "fromJson_string_class", caller);
 
         try {
             return original.invoke(caller, json, classOfT);
-        } catch (IllegalAccessException e){
+        } catch (IllegalAccessException e) {
             throw new RuntimeException(e);// ah, the beauty of Java...
-        } catch (InvocationTargetException e){
+        } catch (InvocationTargetException e) {
+            throw (RuntimeException) e.getCause();
+        }
+    }
+
+    @Replacement(
+            replacingStatic = false,
+            type = ReplacementType.TRACKER,
+            id = "fromJson_string_type",
+            usageFilter = UsageFilter.ANY,
+            category = ReplacementCategory.EXT_0
+    )
+    public static Object fromJson(Object caller, String json, Type typeOfT) {
+        if (caller == null) {
+            throw new NullPointerException();
+        }
+
+        Class<?> klass = (Class<?>) typeOfT;
+        ClassToSchema.registerSchemaIfNeeded(klass);
+        JsonTaint.handlePossibleJsonTaint(json, klass);
+
+        Method original = getOriginal(singleton, "fromJson_string_type", caller);
+
+        try {
+            return original.invoke(caller, json, typeOfT);
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        } catch (InvocationTargetException e) {
+            throw (RuntimeException) e.getCause();
+        }
+    }
+
+    @Replacement(
+            replacingStatic = false,
+            type = ReplacementType.TRACKER,
+            id = "fromJson_reader_class",
+            usageFilter = UsageFilter.ANY,
+            category = ReplacementCategory.EXT_0
+    )
+    public static Object fromJson(Object caller, Reader json, Class<?> classOfT) {
+        if (caller == null) {
+            throw new NullPointerException();
+        }
+
+        ClassToSchema.registerSchemaIfNeeded(classOfT);
+        JsonTaint.handlePossibleJsonTaint(getStringFromReader(json), classOfT);
+
+        Method original = getOriginal(singleton, "fromJson_reader_class", caller);
+
+        try {
+            return original.invoke(caller, json, classOfT);
```

Comment:
this seems wrong, as the stream from `json` has already been red in line 95. see how `InputStream` is handled in `JacksonObjectMapperClassReplacement.readValue`

Reply:
also make sure in unit tests that semantics of function is not alterated